02_bubble/n0npax/main.go

```diff
@@ -0,0 +1,75 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+// partition divides slice into partitions based on pivot high
+func partition(s []int, low int, high int) int {
+	pivot, i := s[high]+1, low-1
+	for j := low; j < high; j++ {
+		if s[j] < pivot {
+			i++
+			s[i], s[j] = s[j], s[i]
+		}
+	}
+	s[i+1], s[high] = s[high], s[i+1]
+	return i + 1
+}
+
+// qsort sorts provided partition. Modifies reference
+func qsort(s []int, lo int, hi int) {
+	if lo < hi {
+		// p - partition vertical pivot
+		p := partition(s, lo, hi)
+		qsort(s, lo, p-1)
+		qsort(s, p+1, hi)
+	}
+
+}
+
+// quicksort returns sorted list(copy) using quicksort
+// quicksort has average compexity O(nlogn)
+func quicksort(s []int) []int {
+	ts := append(s[:0:0], s...)
+	qsort(ts, 0, len(ts)-1)
+	return ts
+}
+
+// insertion sorts list(copy) using insertion
+func insertion(s []int) []int {
+	ts := append(s[:0:0], s...)
+
+	for i := range ts {
+		for i > 0 {
+			if ts[i-1] > ts[i] {
+				ts[i-1], ts[i] = ts[i], ts[i-1]
+			}
+			i--
```

Comment:
Should have commented on this earlier (and no need to change anything).
I've learned by now that the you have implemented `insertion` is a generally accepted solution. 
For completeness, the way I've learned it and Donald Knuth describes it in insertion sort you build up a sorted slice from left to right by finding the right "insertion point" for each each successive element in the already sorted slice on the left.
It's a bit like how you might sort a deck of cards in your hand: you go from left to right, pull out one card at a time and place it in the correct position in the already sorted part.

Here's the solution that we've come up with (after some programming golf):
```go
	for i := 1; i < len(s); i++ {
		var j int
		for j = i; j > 0; j-- {
			if s[j-1] <= s[i] {
				break // We've found the insertion point
			}
		}
		// insert s[i] at s[j] moving the rest up
		tmp := s[i]
		copy(s[j+1:i+1], s[j:i])
		s[j] = tmp
	}
```